### ISSUE TITLE
Using a newer ICU version when compiling the intl extension for PHP

### ISSUE THREAD

**Issue by stof** [travis-ci/travis-ci#3616]
Currently, PHP VMs on Travis are using ICU 4.8. It would be great to have a more uptodate version of ICU in the VMs.

See https://github.com/symfony/symfony/issues/14259 for a discussion in the Symfony community about this issue

**Comment by BanzaiMan** [comment]
@stof I am not familiar with ICU. Where does it come from, and how is it configured in cookbooks? I had a cursory look at the PHP recipes, but was unable to reason about it.

**Comment by stof** [comment]
@banzaiman given that the PHP configuration used by the cookbooks doesn't specify a custom location with the ``--with-icu-dir`` option, PHP will use the system one. And libicu on Ubuntu 12.04 is quite old, given it has not been updated since 2012 (maybe even 2011 depending of the ubuntu update process btw).

**Comment by webmozart** [comment]
See symfony/symfony#14259 for a description about how to compile PHP with a specific version of ICU. ICU (International Components for Unicode) is a C library that is used as foundation of PHP's "intl" extension.

If you compile PHP with the --enable-intl flag (as [done in the cookbook](https://github.com/travis-ci/travis-cookbooks/blob/master/ci_environment/phpbuild/templates/default/default_configure_options.erb)), then PHP will compile intl with the ICU library installed on the system (i.e. Ubuntu's libicu package). You can however use a custom build of ICU by setting the --with-icu-dir option when compiling PHP. Again, see symfony/symfony#14259 for detailed steps.

**Comment by webmozart** [comment]
Some more information: Since ICU 4.8 (used on Travis), ICU was released in the versions 49 (they changed the versioning scheme), 50, 51, 52, 53, 54 and 55. It would be great if Travis used the [latest version 55.1](http://source.icu-project.org/repos/icu/icu/tags/release-55-1/).

**Comment by BanzaiMan** [comment]
Hmm. That'll require updates to the [`libicu` cookbook](https://github.com/travis-ci/travis-cookbooks/tree/master/ci_environment/libicu) as well as [PHP build definition](https://github.com/travis-ci/travis-cookbooks/blob/master/ci_environment/phpbuild/templates/default/default_configure_options.erb).

I'm sorry to say this, but we do not have time to make this in time for tomorrow's build environment update.

We can perhaps manage a PHP-only update before the June update, if anyone wants to work on this.

**Comment by jakzal** [comment]
How's this coming along?

**Comment by jakzal** [comment]
@BanzaiMan seems the libicu cookbook does not exist anymore. Is there anything I could do to help bringing support for newer ICU versions to travis?

**Comment by ADmad** [comment]
We have also started having problems with CakePHP's tests due to outdated ICU lib on travis.

**Comment by BanzaiMan** [comment]
Hello.

I tweaked the build process of our PHP nightly packages to compile ICU (57.1 as of this writing), install and use it. I have not tested it (since I don't have a reasonable PHP app to do so), so if you have one, please test it, and report the results.

ICU installation code: https://github.com/travis-ci/php-src-builder/blob/239e5287f3cabb95d6a25a3c2c5142138844b3cf/bin/install-icu
PHP configure changes: https://github.com/travis-ci/php-src-builder/blob/239e5287f3cabb95d6a25a3c2c5142138844b3cf/.travis.yml#L38

Using nightly:
```yaml
language: php
php:
  - nightly
```

Thanks.

**Comment by ADmad** [comment]
It doesn't work https://travis-ci.org/cakephp/cakephp/jobs/124860744

**Comment by BanzaiMan** [comment]
@ADmad Thanks for testing.

**Comment by BanzaiMan** [comment]
I've put in another tweak to the `nightly` packaging. Could you try it again? Restarting an existing build is fine. My very preliminary test ran successfully. https://travis-ci.org/BanzaiMan/travis_production_test/builds/124927621

**Comment by ADmad** [comment]
Restarted the job. Testsuite runs now and the particular issue we had related to datetime formatting using ICU has been resolved. Though there are other failures on nightly :)

Will the ICU lib be updated for stable php versions?

**Comment by ADmad** [comment]
Btw [this](https://travis-ci.org/cakephp/cakephp/jobs/124860744#L521) test which passes on ubuntu 14.04 which has ICU 52.1 is failing.

**Comment by BanzaiMan** [comment]
@ADmad Thanks for the updates. There are a few moving parts to consider before this build process can be brought to other PHP versions. However, it is safe to say that `nightly` and `7.0snapshot` builds will have compatibility with a newer versions of ICU (57.1, as of this writing, as I've said before). All others, *maybe*.

I am not an expert on what ICU can and cannot do, so I can't say why your test passes with 52.1 and not with 57.1. Does it pass if you build your PHP runtime with ICU 57.1? If you want access to a build VM for live debugging, we are happy to accommodate that for you. Please email support@travis-ci.com.

Thank you!

**Comment by ADmad** [comment]
> Does it pass if you build your PHP runtime with ICU 57.1? If you want access to a build VM for live debugging, we are happy to accommodate that for you.

That test probably would pass using ICU 57.1. But it's not something I want to get into as I personally don't need it :)  As one of the maintainers of CakePHP I was just looking into getting it's testsuite run successfully on travis-ci.

The reason I pointed this out is because ubuntu 14.04 LTS is a pretty popular distro and lot of people would be using it, so maybe having travis VMs which match the libs on it might be more beneficial to users than having the latest ICU version.

I'll leave it to you guys to decide what's best.

Thank you for your quick response on this issue.

**Comment by BanzaiMan** [comment]
We do not yet officially support PHP on Trusty. https://docs.travis-ci.com/user/trusty-ci-environment#PHP

For nightly builds such as `nightly` and `7.0snapshot`, I am inclined to say that having the latest version is important.

For others, if we are to stick with what is available for the Ubuntu release, then we would not be updating ICU at all, since ICU is most up to date at the time the runtimes are compiled.

**Comment by ADmad** [comment]
Eagerly awaiting PHP support on trusty :)

**Comment by jakzal** [comment]
@BanzaiMan I just tried `nightly` with the Symfony test suite. All the ICU related tests in the Intl component are now run [with success]https://travis-ci.org/jakzal/symfony/jobs/124994690#L1938-L1960) (other failure are not related to ICU).

**Comment by BanzaiMan** [comment]
@jakzal Great to hear.

**Comment by BanzaiMan** [comment]
Now, we have a decision to make. Namely, do we want PHP runtimes that has most up-to-date ICU compiled in or not?

As I stated earlier, I think it makes sense to have the latest version for `nightly` and `7.0snapshot`. All other releases, arguments can be made on both sides:

* Have a more recent version compiled, since things are breaking, even if it means linking to a version not found on the system by default.
* Having a more predictable build environment is key, so we should stick to the version available for the release (4.8 on Precise, 52.1 on Trusty).

If the latter is taken, then there is nothing left to do on this issue.

**Comment by ADmad** [comment]
> I think it makes sense to have the latest ICU version for nightly and 7.0snapshot
:+1: 

> Having a more predictable build environment is key, so we should stick to the version available for the release (4.8 on Precise, 52.1 on Trusty).

:+1: 

FYI the just released ubuntu 16.04 LTS has ICU 55.1

**Comment by jakzal** [comment]
Note that it only makes sense to use the latest available ICU data. ICU 55.1 is already outdated, which means it's invalid at least at some part (for example new locale were since introduced). Distros are not doing a good job of keeping it up to date.

From testing perspective I'd love to have both of the following at my disposal:
* the version available for the release (i.e. 4.8 on Precise),
* the latest version available (currently 57.1)

**Comment by ADmad** [comment]
@jakzal Having the latest version as well as the version provided by distro at our disposal would be ideal, though I am not sure how feasible it would be for travis' team to provide that.

**Comment by lorenzo** [comment]
The only think that makes sense to have is the latest version. Testing against outdated data will only give a false perception of test stability and will make it impossible to test projects with the current actual data.

**Comment by jakzal** [comment]
@lorenzo 👍

**Comment by BanzaiMan** [comment]
I think this can be closed. Newer ICU is compiled in recent versions of PHP (7 and up, I think), and it is available on the fly. They will also be available for recent version PHP that are pre-installed on the build images.

If I missed something, do let me know.

**Comment by BanzaiMan** [comment]
This is potentially confusing, and it needs to be clearly documented.

To recap:

Currently, `nightly` and `7.0snapshot` installs PHP runtimes which is compiled against the latest ICU, while no release versions do.

This can cause a problem with apps that rely on newer ICU when a release hits and the apps are deployed to an environment without the newer ICU. It is up to the users to ensure that the newer ICU is available on the environment to which the app is deployed.

**Comment by nicolas-grekas** [comment]
How can use latest ICU? I just tried 7.0snapshot and nightly, both report:
ICU version => 4.8.1.1
ICU Data version => 4.8.1

**Comment by jakzal** [comment]
Indeed, ICU 57.1 might be installed on nigthly, but [the PHP extension is compiled against 4.8.1.1](https://travis-ci.org/jakzal/symfony/jobs/177727234#L1547).

<img width="981" alt="screen shot 2016-11-21 at 22 51 44" src="https://cloud.githubusercontent.com/assets/190447/20503985/19dfa7c0-b03d-11e6-867d-5063369b2ffb.png">

**Comment by BanzaiMan** [comment]
It could very well be that our build configuration is wrong.

**Comment by BanzaiMan** [comment]
I think I located the error whereby the newly installed ICU was ignored by the PHP configuration options. Sorry about that. I'm working on a fix.

**Comment by BanzaiMan** [comment]
Some Precise archives should have it now. 

* https://travis-ci.org/travis-ci/php-src-builder/jobs/177828748#L5521 (master)

Different problems plague other versions.

* https://travis-ci.org/travis-ci/php-src-builder/builds/177833310 (7.0snapshot)
* https://travis-ci.org/travis-ci/php-src-builder/builds/177833340 (7.1snapshot)

* https://travis-ci.org/travis-ci/php-src-builder/jobs/177833439 (pecl can't find redis-2.2.8)

**Comment by stof** [comment]
@BanzaiMan ``pecl info redis`` would give info about the local version of the ``redis`` extension. So it is expected that ``pecl info redis`` fails before the installation. You run this command to determine whether it is already installed, so it should not make the build fail.

**Comment by BanzaiMan** [comment]
@stof Oh, I see. Out of curiosity… Is it possible to query remote repository whether or not a certain version of package is available?

**Comment by stof** [comment]
@BanzaiMan there is ``pecl remote-info``.

But AFAICT, ``pecl info`` is what you expected here, according to the next lines: https://github.com/travis-ci/php-src-builder/blob/a075f3a5cc83b60f24960385b47cefceb3223da8/bin/compile-extensions-common#L10

Your build script was broken by https://github.com/travis-ci/php-src-builder/commit/c09c8651ac2d704dcd406fda8a4acc9cf237ac48 as it now exits when ``pecl info`` fails instead of continuing with a ``return`` variable not set to ``0``.
Your ``pecl_install`` function is not compatible with ``errexit`` currently.

**Comment by BanzaiMan** [comment]
`errexit` is intentional. I'll look into `pecl remote-info`. Thanks!

**Comment by stof** [comment]
@BanzaiMan pecl remote-info will not solve your issue. The call was done to detect whether the extension was already installed locally, and so ``pecl info`` is the right call.
But to make the code compatible with ``errexit``, you will have to change the *way* it is called, as failure exit codes in this call are expected (as you use it as a detection way)

**Comment by BanzaiMan** [comment]
Whee. I'm obviously confused! I'll sit down and think through this again.

Thanks for your patience, everyone.

**Comment by BanzaiMan** [comment]
Trusty jobs remain broken. For example, https://travis-ci.org/travis-ci/php-src-builder/jobs/178002322#L2676. Not quite sure why; I tend to think the configuration is correct, as it works fine on [the Precise jobs](https://travis-ci.org/travis-ci/php-src-builder/jobs/178002326).

**Comment by stof** [comment]
@BanzaiMan it looks like it cannot find the icu symbols there

**Comment by BanzaiMan** [comment]
Right. That indicates the configuration option is somehow wrong, but I haven't figured out why yet.

**Comment by ttnnkkrr** [comment]
so basically symfony is broken? at least on ubuntu 16

**Comment by jakzal** [comment]
@ttnnkkrr no

**Comment by jakzal** [comment]
@BanzaiMan any progress? :)

**Comment by BanzaiMan** [comment]
I think Precise archive should report the correct version now:

https://travis-ci.org/BanzaiMan/travis_production_test/builds/208719173#L163

But the Trusty archive remains unavailable, due to the same compilation error I reported earlier.

If anyone is interested in troubleshooting this further, I can show how to test the build.

**Comment by BanzaiMan** [comment]
Trusty build continues to be a problem, and I am getting very close to just giving up on ICU on Trusty.

If you would like to pitch in on what may be ailing the build, you can try running a build script https://gist.github.com/BanzaiMan/213dfb9c85c33b3b1f1f9c80da619f99#file-build-sh (based on https://travis-ci.org/travis-ci/php-src-builder/jobs/210745527) and try your hand at spotting the problem with the build process.

You can use a recent Trusty Docker image, such as one found in https://hub.docker.com/r/travisci/ci-garnet/tags/.

**Comment by BanzaiMan** [comment]
I've disabled ICU on Trusty nightly builds.

**Comment by jakzal** [comment]
This hasn't been fixed. Can someone reopen please?
